### PR TITLE
PR: Add bulk delete

### LINKED ISSUE #40: Fast clear
Clearing many keys from a single bucket has become a somewhat frequent operation.  Currently we do this by cycling through all of the keys one-by-one and removing them.  Is there a bulk remove-many or remove-bucket operation in boto3 that we can do instead ?

### PR DESCRIPTION
**Pull request by martindurant**
bulk delete takes a list of keys to remove, which must be in the same bucket.
rm, recursive, calls this instead of iterating over individual files, will
be much faster in the case of many files.

Fixes #40  - particularly useful for `clear()` on mappings.

### CODE REVIEW

File: s3fs/core.py [hunk 1]
```diff
@@ -454,6 +454,32 @@ def copy(self, path1, path2):
...
+        """
+        buckets = {split_path(path)[0] for path in pathlist}
+        if len(buckets) > 1:
+            raise ValueError("Bulk delete files should refer to only one bucket")
+        bucket = buckets.pop()
+        if len(pathlist) > 1000:
+            for i in range((len(pathlist) // 1000) + 1):
+                print(i)
+                self.bulk_delete(pathlist[i*1000:(i+1)*1000])
+            return
```
Comment (mrocklin) [inline s3fs/core.py]: You may be interested in [toolz.partition_all](http://toolz.readthedocs.io/en/latest/api.html#toolz.itertoolz.partition_all)

File: s3fs/core.py [hunk 2]
```diff
@@ -454,6 +454,32 @@ def copy(self, path1, path2):
...
+        delete_keys = {'Objects': [{'Key' : split_path(path)[1]} for path
+                                   in pathlist]}
+        try:
+            self.s3.delete_objects(Bucket=bucket, Delete=delete_keys)
+            self.invalidate_cache(bucket)
+        except ClientError:
+            raise IOError('Bulk delete failed')
```
Comment (mrocklin) [inline s3fs/core.py]: I've found it pleasant to pass through the exceptions from boto3.  They tend to have more information about why things failed.  I see two options:

1.  Skip the try-except and let the exception bubble up.  This is simple but can result in users seeing confusing error messages
2.  Include the exception in the error that we produce

**Comment by mrocklin** [comment]
Looks good to me

File: s3fs/core.py (hunk 2 above)
Comment (martindurant) [inline s3fs/core.py]: Was this implemented anywhere that I didn't notice? Should be done in a different PR.

File: s3fs/core.py (hunk 1 above)
Comment (martindurant) [inline s3fs/core.py]: I knew there was a more elegant way to do this, but somehow couldn't wrap it into my head; also, we iterate over the paths list more than once, would be tidier. Still, this works.

File: s3fs/core.py (hunk 2 above)
Comment (mrocklin) [inline s3fs/core.py]: Ah, I had it in local work.  I'll push up something for review.

File: s3fs/core.py (hunk 1 above)
Comment (mrocklin) [inline s3fs/core.py]: Yeah, not suggesting a change.  Just pointing out a function I find I use fairly often.
